extensions/twitch/src/helpers/useTwitchRequest.tsx

```diff
@@ -0,0 +1,149 @@
+/* eslint-disable @typescript-eslint/no-explicit-any -- we should allow `any` as the network response, otherwise we're getting into @tanstack/query territory and it gets complicated fast */
+
+import { LaunchType, Toast, environment, showToast } from "@raycast/api";
+import { useCachedState } from "@raycast/utils";
+import { CACHE_PREFIX, zeroDate } from "./cache";
+import { getHeaders } from "./auth";
+import { useEffect, useState, useRef } from "react";
+import fetch from "node-fetch";
+
+/**
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ * This can be used for structural sharing between JSON values for example.
+ *
+ * Function taken from {@link https://github.com/TanStack/query/blob/main/packages/query-core/src/utils.ts}
+ */
+function replaceEqualDeep<T>(a: unknown, b: T): T;
+function replaceEqualDeep(a: any, b: any): any {
+  if (a === b) {
+    return a;
+  }
+
+  const array = isPlainArray(a) && isPlainArray(b);
+
+  if (array || (isPlainObject(a) && isPlainObject(b))) {
+    const aSize = array ? a.length : Object.keys(a).length;
+    const bItems = array ? b : Object.keys(b);
+    const bSize = bItems.length;
+    const copy: any = array ? [] : {};
+
+    let equalItems = 0;
+
+    for (let i = 0; i < bSize; i++) {
+      const key = array ? i : bItems[i];
+      copy[key] = replaceEqualDeep(a[key], b[key]);
+      if (copy[key] === a[key]) {
+        equalItems++;
+      }
+    }
+
+    return aSize === bSize && equalItems === aSize ? a : copy;
+  }
+
+  return b;
+}
+
+function isPlainArray(value: unknown) {
+  return Array.isArray(value) && value.length === Object.keys(value).length;
+}
+
+/** Copied from {@link https://github.com/jonschlinkert/is-plain-object} */
+function isPlainObject(o: any): o is object {
+  if (!hasObjectPrototype(o)) {
+    return false;
+  }
+
+  // If has no constructor
+  const ctor = o.constructor;
+  if (typeof ctor === "undefined") {
+    return true;
+  }
+
+  // If has modified prototype
+  const prot = ctor.prototype;
+  if (!hasObjectPrototype(prot)) {
+    return false;
+  }
+
+  // If constructor does not have an Object-specific method
+  // eslint-disable-next-line no-prototype-builtins -- we're not modifying it, just checking
+  if (!prot.hasOwnProperty("isPrototypeOf")) {
+    return false;
+  }
+
+  // Most likely a plain Object
+  return true;
+}
+
+function hasObjectPrototype(o: any): boolean {
+  return Object.prototype.toString.call(o) === "[object Object]";
+}
+
+const defaultSelect = (data: any) => data.data;
+
+export function useTwitchRequest<T>({
```

Comment:
Note you could use `useCachedPromise` or `useFetch` from `@raycast/utils` here. That would make some things easier to handle (loading, caching, etc.)

Reply:
I did kinda look into it but they didn't do exactly what I needed (keeping in mind that for some commands (`following`) we need to minimize re-renders as much as possible because they already render *a lot* initially.
- `useFetch` doesn't allow injecting headers as far as I could tell
- `useCachedPromise` I don't remember exactly the issue... I might take a second look at it now that I have a stable functional requirement

Reply:
> useFetch doesn't allow injecting headers as far as I could tell

You can do so in the `options` object: https://developers.raycast.com/utilities/react-hooks/usefetch#arguments.

These hooks are supposed to make both the extension's UX and DX better, so let me know if it doesn't work out to figure out how we can improve them.

Reply:
The issue with `options` is that obtaining the access token is async, and I don't really want it in a state (to minimize renders). Maybe `options` could also take an async function returning the options?